ls1intum/Artemis
# `Programming exercises`: Add possibility to configure Docker run arguments on Jenkins

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
Closes #6042.
See #5971 for the Bamboo implementation.

### Description
Jenkins allows us to pass arguments to the Docker run command as well.
The YAML configuration format has been adapted a bit to be more flexible:
- short arguments like `-v host_path:/container_path` are now possible as well
- the same argument can be passed multiple times, e.g., `--env SOME_VAR=1 --env OTHER_VAR=2`

For Bamboo this change now means that the extra arguments are not passed as
```
--cpus=2
--memory=2g
--pids-limit=1000
--memory-swap=2g
```
but as
```
--cpus
"2"
--memory
"2g"
--pids-limit
"1000"
--memory-swap
"2g"
```
The two variants should be equivalent. So I assume this should not matter for the functionality.
I did not test it yet on Bamboo, since the test servers are not available at the moment.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Create a new programming exercise.
2. The exercise details page you are on now contains a link to the build plan of the base and solution build plans below the task description.
3. Follow one of the links to Jenkins.
4. In the sidebar on the left select the ‘Configure’ entry.
5. Scroll down to the pipeline script definition (very bottom). The script should contain something like
```groovy
agent {
    docker {
        image '[…]'
        label 'docker'
        args '--cpus 2 […]'
    }
}
```


### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseConfiguration.java | 100% of changed code | 100% of changed code |
